### PR TITLE
expression:implement function password.

### PR DESCRIPTION
**Pull request by louishust**

### CODE REVIEW

File: expression/typeinferer.go [hunk 1]
```diff
@@ -427,6 +427,8 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 		tp = x.Args[0].GetType()
 	case ast.RowFunc:
 		tp = x.Args[0].GetType()
+	case ast.PasswordFunc:
+		tp = types.NewFieldType(mysql.TypeVarString)
```
Comment (shenli) [inline expression/typeinferer.go]: Also need to set charset: `chs = v.defaultCharset`.

Comment (shenli) [inline expression/typeinferer.go]: Also need to add a test case in typeinferer_test.go

Comment (louishust) [inline expression/typeinferer.go]: done

**Comment by shenli** [comment]
LGTM
@XuHuaiyu @tiancaiamao PTAL

**Comment by louishust** [comment]
@XuHuaiyu @tiancaiamao PTAL

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -395,7 +395,39 @@ type builtinPasswordSig struct {
...
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	if len(pass) == 0 {
+		d.SetString("")
+		return d, nil
+	}
+
+	// Two stage SHA1 hash of the password
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: add `.` at the end of this comment.

Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: s/stage/stages

**Comment by tiancaiamao** [comment]
I see this Note in the document:

> This function is deprecated as of MySQL 5.7.6 and will be removed in a future MySQL release.

Should TiDB implement it? @shenli

**Comment by louishust** [comment]
@tiancaiamao @shenli  I think password is a useful function

**Comment by shenli** [comment]
@tiancaiamao We need this. Some DBAs still use this to generate password.

**Comment by tiancaiamao** [comment]
But `set password for 'user'@'host' = PASSWORD('xxx')` will not call this function, 
and the algorithm used by TiDB to generate password is not compatible with this method.
@shenli

**Comment by louishust** [comment]
@tiancaiamao  @shenli 
It seems that the TIDB store password not in MySQL way!
So if user migrate data from MySQL using mysqldump, the user's password is changed?

**Comment by tiancaiamao** [comment]
For some historical reasons, TiDB store password not in MySQL way.
Yes, migrate from the `mysql.user` talbe would have some problem.

The risk of this PR is that, if user change password in this way, he can't login any more:

```
update mysql.user set password=PASSWORD('xxx') where user='xxx' and host='xxx';
```

I'll upgrade TiDB to make the stored data compatible with MySQL, so password function 
in this PR can be used.

Please wait for while, thank you for your patience!  @louishust

**Comment by louishust** [comment]
@tiancaiamao I'd like to upgrade TiDB to make the stored data compatible with MySQL

**Comment by tiancaiamao** [comment]
Nice! Feel fun to have a try. 

Basically, you would follow those steps:
1. Add function upgradeToVer10 https://github.com/pingcap/tidb/blob/master/bootstrap.go#L367, which update password to the new format.
2. Change util.EncodePassword, it should do `Sha1Hash([]byte(pwd))` twice. https://github.com/pingcap/tidb/blob/master/util/auth.go#L53
 change util.CalcPassword, remove this line https://github.com/pingcap/tidb/blob/master/util/auth.go#L32
3. Add test

**Comment by louishust** [comment]
@tiancaiamao upgradeToVer10 used to upgrade the old format password stored in TIDB to MySQL format? It seems can not upgrade using simple UPDATE statement cause I have to decode the old format to []byte and then SHA1 the []byte. Add an internal function upgrade_password for the upgrade? Or some other ways?

**Comment by tiancaiamao** [comment]
> upgradeToVer10 used to upgrade the old format password stored in TIDB to MySQL format? 

Yes, you're right.

> It seems can not upgrade using simple UPDATE statement cause I have to decode the old format to []byte and then SHA1 the []byte.

Maybe, but you can have a try.

We don't store the raw password, the stored password in `mysql.user` currently is sha1(password).
it impossible to get the raw password, but... the upgrade just need to change
`sha1(password)` to `sha1(sha1(password))`, i.e. you don't need the raw password, just another sha1 on it.

>  Add an internal function upgrade_password for the upgrade? 

Exactly.



Read the fucking source code would help, including
Auth() in session.go, 
util/auth.go, 
readHandshakeResponse() in server/conn.go which call Auth(),
ConnectionVerification() in privilege/privileges.go.

Here is the password check procedure (maybe miss information in detail):
1. client connect to server
2. server send handshake packet to client,  within the packet there is a random **salt**
3. client send  **salt+sha1(sha1(password))** to server
4. server call Auth() function, check **salt+sha1(sha1pwd)** is the same.

**Comment by louishust** [comment]
this issue is blocked by  #3292 , after #3292 is merged, 
this issue can just use the EncodePassword function in #3292

**Comment by tiancaiamao** [comment]
Please resolve conflict.

**Comment by louishust** [comment]
@tiancaiamao OK

**Comment by tiancaiamao** [comment]
LGTM @XuHuaiyu @shenli

**Comment by shenli** [comment]
LGTM
@louishust Thanks!

**Pull request merged**